File: mobile/lib/src/core/dsa/trie.dart
/// Prefix trie for autocomplete — O(m) insert, O(m + k) prefix lookup.
class TrieNode<T> {
  TrieNode();

  final Map<String, TrieNode<T>> children = {};
  final Set<T> values = {};
  bool terminal = false;
}

class Trie<T> {
  final TrieNode<T> _root = TrieNode<T>();

  void insert(String key, T value) {
    final normalized = key.trim();
    if (normalized.isEmpty) return;
    var node = _root;
    for (final ch in normalized.split('')) {
      node = node.children.putIfAbsent(ch, TrieNode<T>.new);
    }
    node.terminal = true;
    node.values.add(value);
  }

  List<T> searchPrefix(String prefix, {int limit = 10}) {
    final p = prefix.trim();
    if (p.isEmpty || limit < 1) return [];
    var node = _root;
    for (final ch in p.split('')) {
      final next = node.children[ch];
      if (next == null) return [];
      node = next;
    }

    final seen = <T>{};
    final out = <T>[];

    void walk(TrieNode<T> current) {
      if (out.length >= limit) return;
      if (current.terminal) {
        for (final value in current.values) {
          if (seen.add(value)) {
            out.add(value);
            if (out.length >= limit) return;
          }
        }
      }
      for (final child in current.children.values) {
        walk(child);
        if (out.length >= limit) return;
      }
    }

    walk(node);
    return out;
  }
}

File: frontend/lib/dsa/trie.ts
/**
 * Prefix trie for client-side autocomplete — O(m) insert, O(m + k) prefix search.
 */
export class Trie<T> {
  private readonly root: TrieNode<T> = { children: new Map(), values: new Set(), terminal: false };

  insert(key: string, value: T): void {
    const normalized = key.trim();
    if (!normalized) return;
    let node = this.root;
    for (const ch of normalized) {
      let next = node.children.get(ch);
      if (!next) {
        next = { children: new Map(), values: new Set(), terminal: false };
        node.children.set(ch, next);
      }
      node = next;
    }
    node.terminal = true;
    node.values.add(value);
  }

  searchPrefix(prefix: string, limit = 10): T[] {
    const p = prefix.trim();
    if (!p || limit < 1) return [];
    let node = this.root;
    for (const ch of p) {
      const next = node.children.get(ch);
      if (!next) return [];
      node = next;
    }
    const seen = new Set<T>();
    const out: T[] = [];

    const walk = (current: TrieNode<T>) => {
      if (out.length >= limit) return;
      if (current.terminal) {
        for (const v of current.values) {
          if (!seen.has(v)) {
            seen.add(v);
            out.push(v);
            if (out.length >= limit) return;
          }
        }
      }
      for (const child of current.children.values()) {
        walk(child);
        if (out.length >= limit) return;
      }
    };

    walk(node);
    return out;
  }
}

type TrieNode<T> = {
  children: Map<string, TrieNode<T>>;
  values: Set<T>;
  terminal: boolean;
};
